File: backend/scripts/update_market_cap_data.js
const axios = require('axios');
const mysql = require('mysql2/promise');
require('dotenv').config();

// Database configuration
const dbConfig = {
  host: process.env.DB_HOST || 'localhost',
  user: process.env.DB_USER || 'root',
  password: process.env.DB_PASSWORD || '',
  database: process.env.DB_NAME || 'crypto_data',
  charset: 'utf8mb4'
};

// CoinGecko API URLs
const COINGECKO_GLOBAL_API = 'https://api.coingecko.com/api/v3/global';
const COINGECKO_MARKET_CAP_API = 'https://api.coingecko.com/api/v3/global/market_cap_chart';

async function updateMarketCapData() {
  let connection;
  
  try {
    console.log('🔄 Updating Market Cap data...');
    
    // Connect to database
    connection = await mysql.createConnection(dbConfig);
    console.log('✅ Connected to database');

    // Fetch global market data from CoinGecko
    const globalResponse = await axios.get(COINGECKO_GLOBAL_API, {
      timeout: 15000,
      headers: {
        'User-Agent': 'Crypto-Analysis-Dashboard/1.0'
      }
    });

    const globalData = globalResponse.data.data;
    const totalMarketCap = globalData.total_market_cap.usd;
    const marketCapChange24h = globalData.market_cap_change_percentage_24h_usd;

    console.log(`📊 Total Market Cap: $${(totalMarketCap / 1e12).toFixed(2)}T`);
    console.log(`📈 24h Change: ${marketCapChange24h.toFixed(2)}%`);

    // Check if data already exists for today
    const [existingRows] = await connection.execute(
      'SELECT id FROM market_cap_data WHERE date = CURDATE()'
    );

    if (existingRows.length > 0) {
      // Update existing record
      await connection.execute(
        `UPDATE market_cap_data 
         SET total_market_cap = ?, market_cap_change_24h = ?, active_cryptocurrencies = ?, markets = ?, updated_at = NOW()
         WHERE date = CURDATE()`,
        [totalMarketCap, marketCapChange24h, globalData.active_cryptocurrencies, globalData.markets]
      );
      console.log('✅ Updated existing market cap record');
    } else {
      // Insert new record
      await connection.execute(
        `INSERT INTO market_cap_data (date, total_market_cap, market_cap_change_24h, active_cryptocurrencies, markets, created_at, updated_at)
         VALUES (CURDATE(), ?, ?, ?, ?, NOW(), NOW())`,
        [totalMarketCap, marketCapChange24h, globalData.active_cryptocurrencies, globalData.markets]
      );
      console.log('✅ Inserted new market cap record');
    }

    // Update individual token market cap data
    await updateTokenMarketCapData(connection);

    console.log('🎉 Market Cap data update completed successfully!');

  } catch (error) {
    console.error('❌ Error updating market cap data:', error.message);
    if (error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response data:', error.response.data);
    }
  } finally {
    if (connection) {
      await connection.end();
      console.log('🔌 Database connection closed');
    }
  }
}

async function updateTokenMarketCapData(connection) {
  try {
    console.log('🔄 Updating individual token market cap data...');

    // Get list of tokens from database
    const [tokens] = await connection.execute(
      'SELECT symbol, name FROM token_info ORDER BY symbol'
    );

    // Process tokens in batches to avoid rate limiting
    const batchSize = 5;
    for (let i = 0; i < tokens.length; i += batchSize) {
      const batch = tokens.slice(i, i + batchSize);
      
      await Promise.all(batch.map(async (token) => {
        try {
          // Use existing get_market_cap_safe.py logic or CoinGecko API
          const coinId = getCoinGeckoId(token.symbol);
          if (!coinId) {
            console.log(`⚠️ No CoinGecko ID found for ${token.symbol}`);
            return;
          }

          const response = await axios.get(
            `https://api.coingecko.com/api/v3/coins/${coinId}`,
            {
              timeout: 10000,
              headers: {
                'User-Agent': 'Crypto-Analysis-Dashboard/1.0'
              }
            }
          );

          const marketData = response.data.market_data;
          if (!marketData) return;

          const marketCap = marketData.market_cap?.usd;
          const circulatingSupply = marketData.circulating_supply;
          const totalSupply = marketData.total_supply;

          if (marketCap && marketCap > 0) {
            // Check if data exists for today
            const [existingRows] = await connection.execute(
              'SELECT id FROM market_cap_data WHERE symbol = ? AND date = CURDATE()',
              [token.symbol]
            );

            if (existingRows.length > 0) {
              await connection.execute(
                `UPDATE market_cap_data 
                 SET market_cap_usd = ?, circulating_supply = ?, total_supply = ?, updated_at = NOW()
                 WHERE symbol = ? AND date = CURDATE()`,
                [marketCap, circulatingSupply, totalSupply, token.symbol]
              );
            } else {
              await connection.execute(
                `INSERT INTO market_cap_data (symbol, date, market_cap_usd, circulating_supply, total_supply, created_at, updated_at)
                 VALUES (?, CURDATE(), ?, ?, ?, NOW(), NOW())`,
                [token.symbol, marketCap, circulatingSupply, totalSupply]
              );
            }

            console.log(`✅ Updated ${token.symbol}: $${(marketCap / 1e9).toFixed(2)}B`);
          }

          // Rate limiting
          await new Promise(resolve => setTimeout(resolve, 200));

        } catch (error) {
          console.log(`⚠️ Failed to update ${token.symbol}: ${error.message}`);
        }
      }));

      // Wait between batches
      if (i + batchSize < tokens.length) {
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
    }

  } catch (error) {
    console.error('❌ Error updating token market cap data:', error.message);
  }
}

function getCoinGeckoId(symbol) {
  const mapping = {
    'BTC': 'bitcoin',
    'ETH': 'ethereum',
    'BNB': 'binancecoin',
    'SOL': 'solana',
    'XRP': 'ripple',
    'DOGE': 'dogecoin',
    'ADA': 'cardano',
    'AVAX': 'avalanche-2',
    'DOT': 'polkadot',
    'MATIC': 'matic-network',
    'TRX': 'tron',
    'LINK': 'chainlink',
    'UNI': 'uniswap',
    'LTC': 'litecoin',
    'BCH': 'bitcoin-cash',
    'XLM': 'stellar',
    'ATOM': 'cosmos',
    'FIL': 'filecoin',
    'VET': 'vechain',
    'ICP': 'internet-computer'
  };
  
  return mapping[symbol.toUpperCase()];
}

// Run if called directly
if (require.main === module) {
  updateMarketCapData();
}

module.exports = { updateMarketCapData };

File: update_market_cap_data.py
import mysql.connector
from datetime import datetime, timedelta
import requests
import time

# Danh sách 50 token phổ biến nhất
TOP_50_TOKENS = [
    'BTC', 'ETH', 'BNB', 'SOL', 'XRP', 'DOGE', 'TRX', 'ADA', 'AVAX', 'LINK',
    'SUI', 'BCH', 'LTC', 'SHIB', 'TON', 'DOT', 'MNT', 'XMR', 'UNI', 'NEAR',
    'PEPE', 'APT', 'ICP', 'ARB', 'OP', 'INJ', 'HBAR', 'STX', 'VET', 'FIL',
    'MKR', 'ATOM', 'GRT', 'AXS', 'LDO', 'QNT', 'EOS', 'XTZ', 'KSM', 'THETA',
    'BSV', 'DASH', 'ZEC', 'XEM', 'IOTA', 'WAVES', 'ALGO', 'XLM', 'NEO', 'ETC'
]

# Cấu hình database
DB_CONFIG = {
    'host': 'localhost',
    'user': 'root',
    'password': '23082004',
    'database': 'crypto_data',
    'port': 3306
}

# Cấu hình số ngày lùi để lấy dữ liệu (từ ngày hôm nay trở về trước)
DAYS_BACK = 22  # Thay đổi số này để lấy dữ liệu từ bao nhiêu ngày trước

def get_coingecko_id(symbol):
    """Chuyển đổi symbol sang CoinGecko ID"""
    symbol_to_id = {
        'BTC': 'bitcoin',
        'ETH': 'ethereum',
        'BNB': 'binancecoin',
        'SOL': 'solana',
        'XRP': 'ripple',
        'DOGE': 'dogecoin',
        'TRX': 'tron',
        'ADA': 'cardano',
        'AVAX': 'avalanche-2',
        'LINK': 'chainlink',
        'SUI': 'sui',
        'BCH': 'bitcoin-cash',
        'LTC': 'litecoin',
        'SHIB': 'shiba-inu',
        'TON': 'the-open-network',
        'DOT': 'polkadot',
        'MNT': 'mantle',
        'XMR': 'monero',
        'UNI': 'uniswap',
        'NEAR': 'near',
        'PEPE': 'pepe',
        'APT': 'aptos',
        'ICP': 'internet-computer',
        'ARB': 'arbitrum',
        'OP': 'optimism',
        'INJ': 'injective-protocol',
        'HBAR': 'hedera-hashgraph',
        'STX': 'stacks',
        'VET': 'vechain',
        'FIL': 'filecoin',
        'MKR': 'maker',
        'ATOM': 'cosmos',
        'GRT': 'the-graph',
        'AXS': 'axie-infinity',
        'LDO': 'lido-dao',
        'QNT': 'quant-network',
        'EOS': 'eos',
        'XTZ': 'tezos',
        'KSM': 'kusama',
        'THETA': 'theta-token',
        'BSV': 'bitcoin-sv',
        'DASH': 'dash',
        'ZEC': 'zcash',
        'XEM': 'nem',
        'IOTA': 'iota',
        'WAVES': 'waves',
        'ALGO': 'algorand',
        'XLM': 'stellar',
        'NEO': 'neo',
        'ETC': 'ethereum-classic'
    }
    return symbol_to_id.get(symbol.upper())

def fetch_market_cap_data(symbol):
    """Lấy market cap data từ CoinGecko API cho một token"""
    coin_id = get_coingecko_id(symbol)
    if not coin_id:
        return None
    
    try:
        url = f"https://api.coingecko.com/api/v3/coins/{coin_id}"
        response = requests.get(url, timeout=10, params={
            'localization': False,
            'tickers': False,
            'market_data': True,
            'community_data': False,
            'developer_data': False
        })
        
        if response.status_code == 200:
            data = response.json()
            market_data = data.get('market_data', {})
            
            return {
                'market_cap_usd': float(market_data.get('market_cap', {}).get('usd', 0) or 0),
                'circulating_supply': float(market_data.get('circulating_supply', 0) or 0),
                'total_supply': float(market_data.get('total_supply', 0) or 0)
            }
        else:
            print(f"Error fetching data for {symbol}: {response.status_code}")
            return None
    except Exception as e:
        print(f"Error fetching data for {symbol}: {e}")
        return None

def update_database(data, days_back=None):
    """Cập nhật dữ liệu vào MySQL database
    
    Args:
        data: Dữ liệu market cap đã lấy
        days_back: Số ngày lùi để xóa dữ liệu cũ (mặc định: lấy từ DAYS_BACK)
    """
    if days_back is None:
        days_back = DAYS_BACK
    
    connection = None
    cursor = None
    try:
        connection = mysql.connector.connect(**DB_CONFIG)
        cursor = connection.cursor()
        
        # Tính toán ngày bắt đầu
        start_date = (datetime.now() - timedelta(days=days_back)).date()
        today = datetime.now().date()
        current_timestamp = datetime.now()
        
        # Insert dữ liệu mới cho từng token
        insert_query = """
        INSERT INTO market_cap_data (
            symbol, date, market_cap_usd, circulating_supply, total_supply, timestamp
        ) VALUES (
            %s, %s, %s, %s, %s, %s
        )
        ON DUPLICATE KEY UPDATE
            market_cap_usd = VALUES(market_cap_usd),
            circulating_supply = VALUES(circulating_supply),
            total_supply = VALUES(total_supply),
            timestamp = VALUES(timestamp)
        """
        
        values = []
        for item in data:
            values.append((
                item['symbol'],
                today,
                item['market_cap_usd'],
                item['circulating_supply'],
                item['total_supply'],
                current_timestamp
            ))
        
        cursor.executemany(insert_query, values)
        connection.commit()
        
        print(f"Updated {len(values)} market cap records for {today}")
        
    except Exception as e:
        print(f"Error updating database: {e}")
        import traceback
        traceback.print_exc()
        if connection:
            connection.rollback()
    finally:
        if cursor:
            cursor.close()
        if connection and connection.is_connected():
            connection.close()

def main():
    """Hàm chính để cập nhật dữ liệu market cap cho tất cả tokens"""
    print(f"Starting market cap data update for {len(TOP_50_TOKENS)} tokens...")
    print(f"Fetching data for last {DAYS_BACK} days")
    
    all_data = []
    
    for i, symbol in enumerate(TOP_50_TOKENS, 1):
        print(f"\nUpdating {symbol} ({i}/{len(TOP_50_TOKENS)})...")
        
        # Lấy dữ liệu từ CoinGecko API
        market_cap_data = fetch_market_cap_data(symbol)
        
        if market_cap_data:
            all_data.append({
                'symbol': symbol,
                'market_cap_usd': market_cap_data['market_cap_usd'],
                'circulating_supply': market_cap_data['circulating_supply'],
                'total_supply': market_cap_data['total_supply']
            })
            print(f"Fetched: ${market_cap_data['market_cap_usd']:,.0f}")
        else:
            print(f"Failed to fetch data for {symbol}")
        
        # Nghỉ 1 giây để tránh rate limit (CoinGecko free tier: 10-50 calls/minute)
        time.sleep(1)
    
    if all_data:
        # Cập nhật database
        update_database(all_data, days_back=DAYS_BACK)
        print(f"\nMarket cap data update completed! Processed {len(all_data)} tokens")
    else:
        print(f"\nNo data to update")

if __name__ == "__main__":
    main()

